### PR TITLE
test(llmisvc): e2e tests for canary deployment

### PR DESCRIPTION
**Pull request by bartoszmajsak**
**What this PR does / why we need it**:

Adds e2e tests covering the full canary deployment lifecycle for LLMInferenceService groups. These validate traffic splitting, group membership transitions, and edge cases that were previously only covered by the controlled-deployment spike.

**Test cases:**

| Test | What it validates |
|---|---|
| Canary service backend | Baseline 9/1 -> ramp 7/3 -> promote 0/9 under continuous traffic |
| Model name divergence | Different model.name triggers GroupDegraded=MemberDivergence |
| Divergence resolves | Patching model.name clears GroupDegraded |
| Webhook rejection | weight without group rejected at admission (422) |
| Force-stop | `serving.kserve.io/stop` scales to zero, traffic shifts |
| Decommission | Delete member, remaining stays Ready, zero-downtime |
| Leave group | Remove group+weight from spec, member exits group status |
| Three-member group | N>2 members all receive traffic proportionally |
| Late-join | Member joins running group, appears at declared weight |
| Delete at nonzero weight | No route breakage when deleting weight>0 member |
| Rollback | Promote then reverse - traffic returns to original |
| Force-stop route owner | Ownership handoff when route owner is stopped |

**Design:**

- Uses `llm-d-inference-sim` (not real vLLM) for fast, CPU-only execution
- Traffic split verified via `x-inference-pod` response header (pod name prefix matching)
- Gateway convergence polling (`wait_for_healthy_route` with consecutive 2xx) 
- All wait helpers handle transient API errors for CI resilience
- Service backend only (Envoy Gateway compatible) - pool/mixed scenarios as follow-up PR

**Which issue(s) this PR fixes**:

**Feature/Issue validation/testing**:

- [x] All tests pass on kind + Envoy Gateway (884s total)
- [x] Tests pass with controller fix for stopped-member group status (depends on #5847)

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the [documentation](https://github.com/kserve/website)?

```release-note
NONE
```

### CODE REVIEW

File: test/e2e/llmisvc/test_llm_canary_lifecycle.py [hunk 1]
```diff
@@ -0,0 +1,1300 @@
...
+                if c.get("type") == "Ready" and c.get("status") == "True":
+                    return
+        except k8s_client.ApiException:
+            pass
+        time.sleep(5)
+    raise TimeoutError(f"{name} not Ready within {timeout}s")
+
+
+def get_gateway_base_url(api, name, ns):
+    obj = api.get_namespaced_custom_object(
```
Comment (mwaykole) [inline test/e2e/llmisvc/test_llm_canary_lifecycle.py]: This does a one-shot read with no retry. If the controller sets `Ready=True` slightly before populating `status.url` (even by one reconcile tick), this raises immediately. Since every traffic test calls `wait_ready` then immediately hits this, a short poll loop here would prevent spurious failures:

```python
def get_gateway_base_url(api, name, ns, timeout=30):
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        obj = api.get_namespaced_custom_object(...)
        url = obj.get("status", {}).get("url", "")
        if not url:
            addresses = obj.get("status", {}).get("addresses", [])
            if addresses:
                url = addresses[0].get("url", "")
        if url:
            parsed = urlparse(url)
            return f"{parsed.scheme}://{parsed.netloc}"
        time.sleep(2)
    raise TimeoutError(f"No URL in {name} status after {timeout}s")
```

File: test/e2e/llmisvc/test_llm_canary_lifecycle.py [hunk 2]
```diff
@@ -0,0 +1,1300 @@
...
+            wait_for_member_count(api, v1.name, ns, 1)
+
+            driver.mark("settled")
+            time.sleep(15)
+
+            report = driver.stop()
+
+            # Assert across the full mutation window, not just after settlement
+            report.all.assert_no_errors(
+                "decommission: zero errors including transition"
```
Comment (mwaykole) [inline test/e2e/llmisvc/test_llm_canary_lifecycle.py]: This asserts zero errors across the *entire* run including the delete→route-reprogramming window. Other tests in this file (e.g. `test_canary_service_backend`) correctly assert only on stable phases after convergence.

If this is intentional (proving truly zero-downtime member deletion), it's the strongest claim in the file and likely the first test to flake in CI when the gateway has even a brief routing gap. Might be worth a comment noting that expectation, or having a fallback `report.phase("settled").assert_no_errors(...)` ready.

File: test/e2e/llmisvc/test_llm_canary_lifecycle.py [hunk 3]
```diff
@@ -0,0 +1,1300 @@
...
+        )
+
+        # --- Propagation delay ---
+        for mutation, settled in [
+            ("canary_mutation", "canary"),
+            ("promote_mutation", "promote"),
+        ]:
+            delay = report.marks[settled] - report.marks[mutation]
+            logger.info(f"Propagation {mutation} -> {settled}: {delay:.1f}s")
+            assert delay < 30, f"{mutation} -> {settled} took {delay:.1f}s"
```
Comment (mwaykole) [inline test/e2e/llmisvc/test_llm_canary_lifecycle.py]: 30s hard ceiling on gateway propagation could be tight on overloaded CI nodes or when EnvoyGateway xDS push is delayed. Consider making this configurable via env var (e.g. `PROPAGATION_TIMEOUT`) or bumping to 60s as a safe default for CI.

File: test/e2e/llmisvc/test_llm_canary_lifecycle.py [hunk 4]
```diff
@@ -0,0 +1,1300 @@
...
+    for m in (
+        obj.get("status", {}).get("router", {}).get("group", {}).get("members", [])
+    ):
+        if m.get("name") == member:
+            return m
+    return None
+
+
+def wait_for_healthy_route(gateway_url, headers, payload, consecutive=3, timeout=30):
+    """Poll the gateway until we get consecutive 2xx responses.
```
Comment (mwaykole) [inline test/e2e/llmisvc/test_llm_canary_lifecycle.py:410]: nit: `import requests as _requests` inline when `requests` is a standard dep already used by `_driver.py`. Same for `import json as _json` in `test_weight_without_group_rejected`. Moving both to top-level imports would be more consistent with the rest of the file.

File: test/e2e/llmisvc/test_llm_canary_lifecycle.py [hunk 5]
```diff
@@ -0,0 +1,1300 @@
...
+            )
+
+            time.sleep(30)
+            report = driver.stop()
+
+            report.all.assert_no_errors("three-member group")
+
+            for m in [v1, v2, v3]:
+
+                def match(r, prefix=m.name):
```
Comment (mwaykole) [inline test/e2e/llmisvc/test_llm_canary_lifecycle.py]: The `prefix=m.name` default arg correctly captures the value at definition time (avoids the classic loop-variable closure bug). Works fine, but `functools.partial` or a comprehension might make the intent more obvious to future readers who'll do a double-take here.

**Review by mwaykole — COMMENTED** [review]
Nice work. The traffic driver package is well-designed (drift-compensating rate control, mark-based windowing, clean Slice API), test coverage for the canary lifecycle is comprehensive.

No blocking comments — left a few nits and one potential CI flake to watch (`get_gateway_base_url` race, `test_decommission` full-window zero-error assertion). LGTM.

File: test/e2e/llmisvc/test_llm_canary_lifecycle.py (hunk 1 above)
Comment (bartoszmajsak) [reply test/e2e/llmisvc/test_llm_canary_lifecycle.py]: Interesting, I must have been lucky with timing and didn't see flakes. Good catch, adding the poll loop.

File: test/e2e/llmisvc/test_llm_canary_lifecycle.py (hunk 2 above)
Comment (bartoszmajsak) [reply test/e2e/llmisvc/test_llm_canary_lifecycle.py]: Adding a comment documenting the expectation. If it flakes in CI, that's a real signal worth investigating rather than masking.

File: test/e2e/llmisvc/test_llm_canary_lifecycle.py (hunk 3 above)
Comment (bartoszmajsak) [reply test/e2e/llmisvc/test_llm_canary_lifecycle.py]: Makes sense considering that infra might be slower sometimes

File: test/e2e/llmisvc/test_llm_canary_lifecycle.py (hunk 5 above)
Comment (bartoszmajsak) [reply test/e2e/llmisvc/test_llm_canary_lifecycle.py]: Not convinced about `partial` here, seems heavier for this simple case. WDYT?

Comment (mwaykole) [reply test/e2e/llmisvc/test_llm_canary_lifecycle.py]: yeah it will be heavier

**Review by jlost — COMMENTED** [review]
Great idea, very elegantly executed. A couple minor comments:
1. It should integrate the per-test namespacing (just use the fixture param). Some of your manual cleanup will become redundant. 
2. Might it be useful to see more response details for non-200 requests? I could picture getting e.g. a 503 and wanting to take a look at the response body.

**Comment by bartoszmajsak** [comment]
Great suggestions, thanks @jlost!

> 1. It should integrate the per-test namespacing (just use the fixture param). Some of your manual cleanup will become redundant.

This code predates your improvement, fish memory :D Good catch!

> 2. Might it be useful to see more response details for non-200 requests? I could picture getting e.g. a 503 and wanting to take a look at the response body.

I don't see a reason why not? :)

File: test/e2e/llmisvc/traffic/_report.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+    """Single HTTP response captured by the driver."""
+
+    timestamp: float  # time.monotonic()
+    wall_clock: float  # time.time()
+    status: int  # HTTP status code (0 for connection errors)
+    latency: float  # seconds
+    headers: Dict[str, str] = field(default_factory=dict)  # lowercased keys
+    error: Optional[str] = None
+    body: Optional[str] = None  # response body for non-2xx (truncated to 512 chars)
+    body: Optional[str] = None  # response body for non-2xx (truncated to 512 chars)
```
Comment (vivekk16) [inline test/e2e/llmisvc/traffic/_report.py:21]: nit: body is declared twice. Python silently deduplicates it, so no runtime issue.

**Review by vivekk16 — COMMENTED** [review]
nit: The new files are missing the Apache 2.0 copyright header that every other .py file in test/e2e/llmisvc/ has.

And a minor duplication nit under ResponseRecord. Otherwise LGTM!

File: test/e2e/llmisvc/traffic/_report.py (hunk 6 above)
Comment (bartoszmajsak) [reply test/e2e/llmisvc/traffic/_report.py:21]: this is weird copy-paste issue, will address as follow-up. not sure what happened here.

**Comment by bartoszmajsak** [comment]
> nit: The new files are missing the Apache 2.0 copyright header that every other .py file in test/e2e/llmisvc/ has.

coming as follow-up. sharp eye @vivekk16 !

**Comment by bartoszmajsak** [comment]
/rerun-failed

ARGH!

**Comment by bartoszmajsak** [comment]
> > nit: The new files are missing the Apache 2.0 copyright header that every other .py file in test/e2e/llmisvc/ has.
> 
> coming as follow-up. sharp eye @vivekk16 !

@vivekk16 it was a bit more involved than just these new net files. I hope that's the right fix - #5864

File: test/e2e/llmisvc/traffic/_report.py (hunk 6 above)
Comment (bartoszmajsak) [reply test/e2e/llmisvc/traffic/_report.py:21]: #5865
